File: src/main/java/main/tut14/PerspectiveInterpolation.java

package main.tut14;

import com.jogamp.newt.event.KeyEvent;
import com.jogamp.opengl.GL3;
import com.jogamp.opengl.GLAutoDrawable;
import com.jogamp.opengl.GLRunnable;
import main.framework.Framework;
import main.framework.component.Mesh;
import org.xml.sax.SAXException;
import uno.glm.MatrixStack;

import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.logging.Level;
import java.util.logging.Logger;

import static com.jogamp.opengl.GL2ES3.GL_COLOR;
import static com.jogamp.opengl.GL2ES3.GL_DEPTH;
import static uno.glsl.UtilKt.programOf;

/**
 * @author elect
 */
public class PerspectiveInterpolation extends Framework {

    public static void main(String[] args) {
        new PerspectiveInterpolation("Tutorial 14 - Perspective Interpolation");
    }

    private ProgramData smoothInterp, linearInterp;

    private Mesh realHallway, fauxHallway;

    private boolean useFakeHallway = false, useSmoothInterpolation = true;

    public PerspectiveInterpolation(String title) {
        super(title);
    }

    @Override
    public void init(GL3 gl) {

        initializePrograms(gl);

        try {
            realHallway = new Mesh(gl, getClass(), "tut14/RealHallway.xml");
            fauxHallway = new Mesh(gl, getClass(), "tut14/FauxHallway.xml");
        } catch (ParserConfigurationException | SAXException | IOException | URISyntaxException ex) {
            Logger.getLogger(PerspectiveInterpolation.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    private void initializePrograms(GL3 gl) {

        smoothInterp = new ProgramData(gl, "smooth-vertex-colors");
        linearInterp = new ProgramData(gl, "no-correct-vertex-colors");

        float zNear = 1.0f, zFar = 1_000f;
        MatrixStack persMatrix = new MatrixStack();
        persMatrix.perspective(60.0f, 1.0f, zNear, zFar);

        gl.glUseProgram(smoothInterp.theProgram);
        gl.glUniformMatrix4fv(smoothInterp.cameraToClipMatrixUnif, 1, false, persMatrix.top().to(matBuffer));
        gl.glUseProgram(linearInterp.theProgram);
        gl.glUniformMatrix4fv(linearInterp.cameraToClipMatrixUnif, 1, false, matBuffer);
        gl.glUseProgram(0);
    }

    @Override
    public void display(GL3 gl) {

        gl.glClearBufferfv(GL_COLOR, 0, clearColor.put(0, 0.0f).put(1, 0.0f).put(2, 0.0f).put(3, 0.0f));
        gl.glClearBufferfv(GL_DEPTH, 0, clearDepth.put(0, 1.0f));

        if (useSmoothInterpolation)
            gl.glUseProgram(smoothInterp.theProgram);
        else
            gl.glUseProgram(linearInterp.theProgram);

        if (useFakeHallway)
            fauxHallway.render(gl);
        else
            realHallway.render(gl);

        gl.glUseProgram(0);
    }

    @Override
    public void reshape(GL3 gl, int w, int h) {
        gl.glViewport(0, 0, w, h);
    }

    @Override
    public void keyPressed(KeyEvent e) {

        switch (e.getKeyCode()) {

            case KeyEvent.VK_ESCAPE:
                quit();
                break;

            case KeyEvent.VK_S:
                useFakeHallway = !useFakeHallway;
                System.out.println(useFakeHallway ? "Fake Hallway." : "Real Hallway.");
                break;
            case KeyEvent.VK_P:
                useSmoothInterpolation = !useSmoothInterpolation;
                System.out.println(useSmoothInterpolation ? "Perspective correct interpolation." : "Just linear interpolation");
                break;
            case KeyEvent.VK_SPACE:
                window.invoke(false, new GLRunnable() {
                    @Override
                    public boolean run(GLAutoDrawable drawable) {

                        GL3 gl = drawable.getGL().getGL3();

                        realHallway.dispose(gl);
                        fauxHallway.dispose(gl);

                        try {
                            realHallway = new Mesh(gl, getClass(), "tut14/RealHallway.xml");
                            fauxHallway = new Mesh(gl, getClass(), "tut14/FauxHallway.xml");
                        } catch (ParserConfigurationException | SAXException | IOException | URISyntaxException ex) {
                            Logger.getLogger(PerspectiveInterpolation.class.getName()).log(Level.SEVERE, null, ex);
                        }

                        return false;
                    }
                });
                break;
        }
    }

    @Override
    public void end(GL3 gl) {

        gl.glDeleteProgram(smoothInterp.theProgram);
        gl.glDeleteProgram(linearInterp.theProgram);

        fauxHallway.dispose(gl);
        realHallway.dispose(gl);
    }

    private class ProgramData {

        public int theProgram;

        public int cameraToClipMatrixUnif;

        public ProgramData(GL3 gl, String shader) {

            theProgram = programOf(gl, getClass(), "tut14", shader + ".vert", shader + ".frag");

            cameraToClipMatrixUnif = gl.glGetUniformLocation(theProgram, "cameraToClipMatrix");
        }
    }
}

File: src/main/kotlin/main/tut14/perspectiveInterpolation.kt
package main.tut14

import com.jogamp.newt.event.KeyEvent
import com.jogamp.opengl.GL2ES3.GL_COLOR
import com.jogamp.opengl.GL2ES3.GL_DEPTH
import com.jogamp.opengl.GL3
import glNext.*
import main.framework.Framework
import main.framework.component.Mesh
import uno.buffer.put
import uno.gl.gl3
import uno.glm.MatrixStack
import uno.glsl.programOf

/**
 * Created by elect on 28/03/17.
 */

fun main(args: Array<String>) {
    PerspectiveInterpolation_().setup("Tutorial 14 - Perspective Interpolation")
}

class PerspectiveInterpolation_() : Framework() {

    lateinit var smoothInterp: ProgramData
    lateinit var linearInterp: ProgramData

    lateinit var realHallway: Mesh
    lateinit var fauxHallway: Mesh

    var useFakeHallway = false
    var useSmoothInterpolation = true

    override fun init(gl: GL3) {

        initializePrograms(gl)

        realHallway = Mesh(gl, javaClass, "tut14/RealHallway.xml")
        fauxHallway = Mesh(gl, javaClass, "tut14/FauxHallway.xml")
    }

    fun initializePrograms(gl: GL3) = with(gl) {

        smoothInterp = ProgramData(gl, "smooth-vertex-colors")
        linearInterp = ProgramData(gl, "no-correct-vertex-colors")

        val zNear = 1.0f
        val zFar = 1_000f
        val persMatrix = MatrixStack()
        persMatrix.perspective(60.0f, 1.0f, zNear, zFar)

        glUseProgram(smoothInterp.theProgram)
        glUniformMatrix4f(smoothInterp.cameraToClipMatrixUnif, persMatrix.top())
        glUseProgram(linearInterp.theProgram)
        glUniformMatrix4f(linearInterp.cameraToClipMatrixUnif, persMatrix.top())
        glUseProgram()
    }

    override fun display(gl: GL3) = with(gl) {

        glClearBufferf(GL_COLOR, 0.0f)
        glClearBufferf(GL_DEPTH)

        if (useSmoothInterpolation)
            glUseProgram(smoothInterp.theProgram)
        else
            glUseProgram(linearInterp.theProgram)

        if (useFakeHallway)
            fauxHallway.render(gl)
        else
            realHallway.render(gl)

        glUseProgram(0)
    }

    override fun reshape(gl: GL3, w: Int, h: Int) = with(gl) {
        glViewport(w, h)
    }

    override fun keyPressed(e: KeyEvent) {

        when (e.keyCode) {

            KeyEvent.VK_ESCAPE -> quit()

            KeyEvent.VK_S -> {
                useFakeHallway = !useFakeHallway
                println(if (useFakeHallway) "Fake Hallway." else "Real Hallway.")
            }
            KeyEvent.VK_P -> {
                useSmoothInterpolation = !useSmoothInterpolation
                println(if (useSmoothInterpolation) "Perspective correct interpolation." else "Just linear interpolation")
            }
            KeyEvent.VK_SPACE -> window.gl3 {

                realHallway.dispose(this)
                fauxHallway.dispose(this)

                realHallway = Mesh(this, javaClass, "tut14/RealHallway.xml")
                fauxHallway = Mesh(this, javaClass, "tut14/FauxHallway.xml")
            }
        }
    }

    override fun end(gl: GL3) = with(gl) {

        glDeletePrograms(smoothInterp.theProgram, linearInterp.theProgram)

        fauxHallway.dispose(gl)
        realHallway.dispose(gl)
    }

    class ProgramData(gl: GL3, shader: String) {
        val theProgram = programOf(gl, PerspectiveInterpolation_::class.java, "tut14", shader + ".vert", shader + ".frag")
        val cameraToClipMatrixUnif = gl.glGetUniformLocation(theProgram, "cameraToClipMatrix")
    }
}